File: Source/JavaScript/engine/for_evaluateFlowArrangement/when_evaluating_against_the_shared_fixture_corpus.ts
// Copyright (c) Cratis. All rights reserved.
// Licensed under the MIT license. See LICENSE file in the project root for full license information.

import { readFileSync } from 'node:fs';
import { join } from 'node:path';
import { ExternalComponent, FlowArrangement, FlowLeaf, FlowOverride, HeightSizeClass, HorizontalAlignment, SizeClass, VerticalAlignment, Visibility, WidthSizeClass } from '@cratis/scene.model';
import { evaluateFlowArrangement } from '../index';

interface FixtureOverride {
    width: WidthSizeClass | null;
    height: HeightSizeClass | null;
    tag: string;
}

interface FixtureCase {
    name: string;
    rootTag: string;
    overrides: FixtureOverride[];
    sizeClass: SizeClass;
    expectedTag: string;
}

interface FixtureCorpus {
    flowCases: FixtureCase[];
}

const manifestPath = join(import.meta.dirname, '..', '..', '..', '..', 'layout-evaluation-fixtures.json');
const corpus = JSON.parse(readFileSync(manifestPath, 'utf-8')) as FixtureCorpus;

function leaf(tag: string): FlowLeaf {
    const content: ExternalComponent = {
        id: tag,
        name: tag,
        properties: {},
        visibility: Visibility.Visible,
        isEnabled: true,
        opacity: 1,
        size: {},
        zIndex: 0,
        minimumSize: {},
        maximumSize: {},
        margin: { left: 0, top: 0, right: 0, bottom: 0 },
        horizontalAlignment: HorizontalAlignment.Stretch,
        verticalAlignment: VerticalAlignment.Stretch,
        componentName: 'core:text',
        slots: {},
    };
    return { content };
}

describe('when evaluating against the shared fixture corpus', () => {
    for (const fixtureCase of corpus.flowCases) {
        it(`should match the expected tag for "${fixtureCase.name}"`, () => {
            const arrangement: FlowArrangement = {
                root: leaf(fixtureCase.rootTag),
                overrides: fixtureCase.overrides.map((override): FlowOverride => ({
                    width: override.width ?? undefined,
                    height: override.height ?? undefined,
                    root: leaf(override.tag),
                })),
            };

            const result = evaluateFlowArrangement(arrangement, fixtureCase.sizeClass) as FlowLeaf;
            (result.content as ExternalComponent).id.should.equal(fixtureCase.expectedTag);
        });
    }
});

File: Source/DotNET/Engine.Specs/for_FlowArrangementEvaluator/when_evaluating_against_the_shared_fixture_corpus.cs
// Copyright (c) Cratis. All rights reserved.
// Licensed under the MIT license. See LICENSE file in the project root for full license information.

using System.Text.Json;
using Cratis.Scene.Engine.Layouts;
using Cratis.Scene.Model.Elements;
using Cratis.Scene.Model.Layouts;
using Cratis.Scene.Model.SizeClasses;

namespace Cratis.Scene.Engine.for_FlowArrangementEvaluator;

public class when_evaluating_against_the_shared_fixture_corpus : Specification
{
    record FixtureCase(string Name, FlowArrangement Arrangement, SizeClass SizeClass, string ExpectedTag);

    List<FixtureCase> _cases = null!;
    List<(FixtureCase Case, string ActualTag)> _results = null!;

    void Establish()
    {
        var manifestPath = Path.Combine(FindRepositoryRoot(), "layout-evaluation-fixtures.json");
        using var document = JsonDocument.Parse(File.ReadAllText(manifestPath));

        _cases = document.RootElement.GetProperty("flowCases").EnumerateArray().Select(ToFixtureCase).ToList();
    }

    void Because() => _results = [.. _cases.Select(fixtureCase => (fixtureCase, TagOf(FlowArrangementEvaluator.Evaluate(fixtureCase.Arrangement, fixtureCase.SizeClass))))];

    [Fact]
    void should_match_the_expected_tag_for_every_case()
    {
        foreach (var (fixtureCase, actualTag) in _results)
        {
            (fixtureCase.Name, actualTag).ShouldEqual((fixtureCase.Name, fixtureCase.ExpectedTag));
        }
    }

    static FixtureCase ToFixtureCase(JsonElement element)
    {
        var name = element.GetProperty("name").GetString()!;
        var root = Leaf(element.GetProperty("rootTag").GetString()!);

        var overrides = element.GetProperty("overrides").EnumerateArray()
            .Select(overrideElement => new FlowOverride(
                ParseWidth(overrideElement),
                ParseHeight(overrideElement),
                Leaf(overrideElement.GetProperty("tag").GetString()!)))
            .ToList();

        var sizeClassElement = element.GetProperty("sizeClass");
        var sizeClass = new SizeClass(
            Enum.Parse<WidthSizeClass>(sizeClassElement.GetProperty("width").GetString()!),
            Enum.Parse<HeightSizeClass>(sizeClassElement.GetProperty("height").GetString()!));

        var expectedTag = element.GetProperty("expectedTag").GetString()!;

        return new(name, new FlowArrangement(root, overrides), sizeClass, expectedTag);
    }

    static WidthSizeClass? ParseWidth(JsonElement overrideElement)
    {
        var widthElement = overrideElement.GetProperty("width");
        return widthElement.ValueKind == JsonValueKind.Null ? null : Enum.Parse<WidthSizeClass>(widthElement.GetString()!);
    }

    static HeightSizeClass? ParseHeight(JsonElement overrideElement)
    {
        var heightElement = overrideElement.GetProperty("height");
        return heightElement.ValueKind == JsonValueKind.Null ? null : Enum.Parse<HeightSizeClass>(heightElement.GetString()!);
    }

    static FlowLeaf Leaf(string tag) => new(new ExternalComponent { Id = tag, Name = tag, ComponentName = "core:text" });

    static string TagOf(FlowNode node) => ((FlowLeaf)node).Content.Id;

    static string FindRepositoryRoot()
    {
        var directory = new DirectoryInfo(AppContext.BaseDirectory);
        while (directory is not null && !File.Exists(Path.Combine(directory.FullName, "Scene.slnx")))
        {
            directory = directory.Parent;
        }

        return directory?.FullName ?? throw new DirectoryNotFoundException("Could not locate the repository root (Scene.slnx) above " + AppContext.BaseDirectory);
    }
}
